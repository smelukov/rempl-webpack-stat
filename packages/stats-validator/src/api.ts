import { ViewConfig } from '@statoscope/types';
import { TestEntry, Details } from './test-entry';

export type RuleResult<TData> = {
  data: TData;
  view: ViewConfig<TData, unknown>;
};

export type Storage = TestEntry[];

export type APIFnOptions = {
  filename?: string;
  compilation?: string;
  details?: Details;
};

export type API = {
  error(message: string, filenameOrOptions?: string | APIFnOptions): void;
  warn(message: string, filenameOrOptions?: string | APIFnOptions): void;
  info(message: string, filenameOrOptions?: string | APIFnOptions): void;
  hasErrors(): boolean;
  getInfoTotal(): number;
  getWarnTotal(): number;
  getErrorTotal(): number;
  getStorage(): Storage;
};

export type MakeAPIParams = {
  warnAsError: boolean;
};

export function makeAPI(params: MakeAPIParams): API {
  const storage: Storage = [];
  let hasErrors = false;
  let errors = 0;
  let warnings = 0;
  let infos = 0;
  return {
    hasErrors(): boolean {
      return hasErrors;
    },
    getInfoTotal(): number {
      return infos;
    },
    getWarnTotal(): number {
      return warnings;
    },
    getErrorTotal(): number {
      return errors;
    },
    getStorage(): Storage {
      return storage;
    },
    warn(message, filenameOrOptions): void {
      if (params.warnAsError) {
        hasErrors = true;
      }

      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      storage.push({
        type: 'warn',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        message,
      });
      warnings++;
    },
    error(message, filenameOrOptions): void {
      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      hasErrors = true;
      storage.push({
        type: 'error',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        message,
      });
      errors++;
    },
    info(message, filenameOrOptions): void {
      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      storage.push({
        type: 'info',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        message,
      });
      infos++;
    },
  };
}
