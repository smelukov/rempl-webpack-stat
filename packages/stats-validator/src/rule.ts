import { API } from './api';

export type Rule<TParams, TInput = unknown> = (
  params: TParams,
  data: TInput,
  api: API
) => void;
