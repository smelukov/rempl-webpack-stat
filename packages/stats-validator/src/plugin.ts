import { Rule } from './rule';

export type InputFile = {
  name: string;
  data: any;
};

export type PrepareFn<TPrepared> = (files: InputFile[]) => TPrepared;
export type PluginDescription<TPrepared> = {
  prepare?: PrepareFn<TPrepared>;
  rules?: Record<string, Rule<any, any>>;
};

export type PluginFn<TPrepared> = () => PluginDescription<TPrepared>;
