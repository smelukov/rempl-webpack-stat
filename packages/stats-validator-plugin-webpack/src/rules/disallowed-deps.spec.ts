import { makeAPI } from '@statoscope/stats-validator/dist/api';
import plugin from '../';
import statsV5 from '../../../../test/bundles/simple/stats-prod.json';
import rule from './disallowed-deps';

describe('package', () => {
  test('matches', () => {
    const pluginInstance = plugin();
    const prepared = pluginInstance.prepare!([{ name: 'stats.json', data: statsV5 }]);
    const api = makeAPI({ warnAsError: false });

    rule(['foo'], prepared, api);
    rule(['foo@^1.0.0 || ^2.0.0'], prepared, api);

    rule([{ type: 'package', name: 'foo' }], prepared, api);
    rule([{ type: 'package', name: /^foo/ }], prepared, api);
    rule([{ type: 'package', name: 'foo', version: '^1.0.0 || ^2.0.0' }], prepared, api);

    expect(api.getStorage()).toMatchSnapshot();
  });

  test('not matches', () => {
    const pluginInstance = plugin();
    const prepared = pluginInstance.prepare!([{ name: 'stats.json', data: statsV5 }]);
    const api = makeAPI({ warnAsError: false });

    rule(['foo@^2.0.0'], prepared, api);

    rule([{ type: 'package', name: /^fo$/ }], prepared, api);
    rule([{ type: 'package', name: 'foo', version: '^2.0.0' }], prepared, api);

    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('module', () => {
  test('matches', () => {
    const pluginInstance = plugin();
    const prepared = pluginInstance.prepare!([{ name: 'stats.json', data: statsV5 }]);
    const api = makeAPI({ warnAsError: false });

    rule(['./src/index.ts'], prepared, api);
    rule([/^\.\/src/], prepared, api);

    rule([{ type: 'module', name: './src/index.ts' }], prepared, api);
    rule([{ type: 'module', name: /^\.\/src/ }], prepared, api);

    expect(api.getStorage()).toMatchSnapshot();
  });

  test('not matches', () => {
    const pluginInstance = plugin();
    const prepared = pluginInstance.prepare!([{ name: 'stats.json', data: statsV5 }]);
    const api = makeAPI({ warnAsError: false });

    rule(['./srF/index.ts'], prepared, api);
    rule([/^\.\/srF/], prepared, api);

    rule([{ type: 'module', name: './srF/index.ts' }], prepared, api);
    rule([{ type: 'module', name: /^\.\/srF/ }], prepared, api);

    expect(api.getStorage()).toMatchSnapshot();
  });
});
