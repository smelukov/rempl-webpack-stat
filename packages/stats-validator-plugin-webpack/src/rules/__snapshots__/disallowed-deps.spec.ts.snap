// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module matches 1`] = `
Array [
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index.ts + 3 modules should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index.ts
",
    "filename": "stats.json",
    "message": "./src/sub-module.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index2.ts + 1 modules should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index2.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index.ts
",
    "filename": "stats.json",
    "message": "./src/sub-module.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index2.ts + 1 modules
",
    "filename": "stats.json",
    "message": "./src/index2.css should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index2.ts + 1 modules
",
    "filename": "stats.json",
    "message": "./src/test.txt should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index2.ts + 1 modules
",
    "filename": "stats.json",
    "message": "./src/dyn.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└─┬ ./src/index2.css
  └── ./src/index2.ts
",
    "filename": "stats.json",
    "message": "./src/statoscope.png should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index.ts + 3 modules should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index.ts
",
    "filename": "stats.json",
    "message": "./src/sub-module.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index2.ts + 1 modules should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "No module reasons (seems like this is an entrypoint)
",
    "filename": "stats.json",
    "message": "./src/index2.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index.ts
",
    "filename": "stats.json",
    "message": "./src/sub-module.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index2.ts + 1 modules
",
    "filename": "stats.json",
    "message": "./src/index2.css should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index2.ts + 1 modules
",
    "filename": "stats.json",
    "message": "./src/test.txt should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└── ./src/index2.ts + 1 modules
",
    "filename": "stats.json",
    "message": "./src/dyn.ts should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Module reasons:
└─┬ ./src/index2.css
  └── ./src/index2.ts
",
    "filename": "stats.json",
    "message": "./src/statoscope.png should not be used",
    "type": "error",
  },
]
`;

exports[`module not matches 1`] = `Array []`;

exports[`package matches 1`] = `
Array [
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Instance reasons [node_modules/foo]:
└─┬ ./src/sub-module.ts
  └── ./src/index.ts
",
    "filename": "stats.json",
    "message": "foo@1.0.0 should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Instance reasons [node_modules/foo]:
└─┬ ./src/sub-module.ts
  └── ./src/index.ts
",
    "filename": "stats.json",
    "message": "foo@1.0.0 should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Instance reasons [node_modules/foo]:
└─┬ ./src/sub-module.ts
  └── ./src/index.ts
",
    "filename": "stats.json",
    "message": "foo@1.0.0 should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Instance reasons [node_modules/foo]:
└─┬ ./src/sub-module.ts
  └── ./src/index.ts
",
    "filename": "stats.json",
    "message": "foo@1.0.0 should not be used",
    "type": "error",
  },
  Object {
    "compilation": "d395b2ff19d3d6f0efe3",
    "details": "Instance reasons [node_modules/foo]:
└─┬ ./src/sub-module.ts
  └── ./src/index.ts
",
    "filename": "stats.json",
    "message": "foo@1.0.0 should not be used",
    "type": "error",
  },
]
`;

exports[`package not matches 1`] = `Array []`;
